src/rez/utils/request_directives.py

```diff
@@ -0,0 +1,280 @@
+
+from rez.vendor.schema.schema import Schema, Optional, Use, And
+from rez.vendor.version.version import VersionRange
+from rez.vendor.version.requirement import Requirement
+from rez.vendor.version.util import dewildcard
+from rez.utils.formatting import PackageRequest
+from copy import copy
+import inspect
+import sys
+
+
+# directives
+#
+
+class DirectiveBase(object):
+    """Base class of directive request handler"""
+    @classmethod
+    def name(cls):
+        """Return the name of the directive"""
+        raise NotImplementedError
+
+    def parse(self, arg_string):
+        """Parse arguments from directive syntax string"""
+        raise NotImplementedError
+
+    def to_string(self, args):
+        """Format arguments to directive syntax string"""
+        raise NotImplementedError
+
+    def process(self, range_, version, rank=None):
```

Comment:
I'm not sure what the idea is here. We have the abstract notion of a 'directive', which is anything added to a pkg request via `//`, but then `rank` is included here in the process method, even though the idea of rank is specific to the harden directive. It seems like there is the abstraction here but at the same time it's too specific to be useful for anything other than the harden directive?

Should it not work something like:
* directive instances are parsed out of a package.py, then somehow associated so that they can be applied back onto the package later on
* the package is updated to replace the original package request with whatever the directive says to do (so harden for eg would replace a req like foo-1.* with foo-1)
* the build resolve occurs
* the package is then updated with the directives applied (so foo-1.* becomes foo-1.4 for eg)

There needs to be some explanation like what I've just done here, giving an overview of what's going on here and how the process works as a whole. I'm having trouble assembling the bigger picture from reading the code at the moment.

Reply:
I'm also confused too because it seems like the old wilcard-based syntax, and the harden syntax, aren't quite the same thing.

What I mean is:
* foo-1.2+<2//harden(2) might expand to foo-1.2.3 (is that right?)
* but foo-1.2.*+<2 might expand to foo-1.2.3+<2 (?)

They're fairly different things, as one (harden) converts to a specific resolved version, whereas the other (wildcards) can retain the original request range. Does it work like this currently? I am not clear on what behaviour to expect.

Reply:
> Does it work like this currently?

Currently this kind of *multi-rank* version range is not supported and should fall back to old-style expansion.
Because, just like the example you have above, I thought the result of `foo-1.2+<2//harden(2)` could be something like `foo-1.2+<2.0.0`, but wasn't sure at the time being.

Reply:
Ah right. Yeah I think this is a little trickier than I expected.

I think it only makes sense for harden() to resolve to a specific version and nothing else. Eg foo-1//harden(2) -> foo-1.3, foo-1.3+<2//harden(3) -> foo.1.3.5. You can't be guaranteed to non-ambiguously 'harden' any version range, and even if you could do it somehow, it wouldn't be intuitive. For that I think wildcard syntax is _required_. So I actually think we need both. Eg, you can do this (but not with harden directive): foo-1.*+<2 -> foo-1.3+<2. But other times you might actually want foo-1.*+<2 -> foo-1.3, and that's when you need harden. Maybe this does actually call for two different Harden directive classes.

I don't think we should ever fall back to pre-build expansion, mixing that up with post-build expansion is just going to be confusing and could lead to bugs that are difficult to track down.